File: packages/cli/src/templates/trellis/scripts/common/adapter_middleware.py
"""Stage 6 Event Bridge + Middleware Provider projection.

Kernel extras remain authoritative. This module mirrors the contract for
hooks and tests. It does not write a second store or invent Command ops.
"""

from __future__ import annotations

from typing import Any

STAGE6_SOURCE = "stage6-adapter-middleware"
STAGE6_SCHEMA_VERSION = 1
SMART_SEARCH_PROVIDER = "smart-search"
EXTERNAL_KNOWLEDGE_CAPABILITY = "external-knowledge"
RETRIEVAL_INTENTS = ("exact", "semantic", "structural", "external")
OPTIONAL_CODE_INTEL_PROVIDERS = ("codegraph", "fast-context")
SHIPPED_MIDDLEWARE_PROVIDERS = (
    "smart-search",
    "codegraph",
    "fast-context",
    "chrome-cdp",
    "playwright",
    "github",
    "cursor-ide-browser",
)
DEFAULT_REQUIRED_MIDDLEWARE_PROVIDERS = (SMART_SEARCH_PROVIDER,)
SHIPPED_PROVIDER_CAPABILITY = {
    "smart-search": EXTERNAL_KNOWLEDGE_CAPABILITY,
    "codegraph": "structural",
    "fast-context": "semantic",
    "chrome-cdp": "browser-session",
    "playwright": "browser-automation",
    "github": "vcs-host",
    "cursor-ide-browser": "ide-browser",
}
SHIPPED_PROVIDER_PROBE = {
    "smart-search": "cli",
    "codegraph": "mcp",
    "fast-context": "mcp",
    "chrome-cdp": "cli",
    "playwright": "mcp",
    "github": "mcp",
    "cursor-ide-browser": "host",
}


class AdapterMiddlewareError(RuntimeError):
    """Event Bridge / Middleware contract rejected a request."""


PROFILE_BASELINE_MODULES = (
    "intake-basic",
    "define-basic",
    "approval-personal",
    "execute-agent",
    "verify-basic",
    "close-basic",
    "context-progressive",
    "observability-local",
)

# Hook subscriptions filled/cleared with Profile activation. Unsubscribed
# hooks stay no-op and must not crash.
MODULE_HOOK_SUBSCRIPTIONS: dict[str, tuple[dict[str, str], ...]] = {
    "context-progressive": (
        {
            "event": "sessionStart",
            "module": "context-progressive",
            "hook": "session-start.py",
        },
        {
            "event": "beforeShellExecution",
            "module": "context-progressive",
            "hook": "inject-shell-session-context.py",
        },
    ),
    "worker-orchestration": (
        {
            "event": "preToolUse",
            "module": "worker-orchestration",
            "hook": "inject-subagent-context.py",
        },
    ),
    "retrieval-extended": (
        {
            "event": "beforeSubmitPrompt",
            "module": "retrieval-extended",
            "hook": "inject-retrieval-plan.py",
        },
        {
            "event": "stop",
            "module": "retrieval-extended",
            "hook": "research-end-retrieval-pack.py",
        },
    ),
}


def default_event_bridge() -> dict[str, Any]:
    return {
        "schema_version": STAGE6_SCHEMA_VERSION,
        "source": STAGE6_SOURCE,
        "subscriptions": [],
    }


def _unique_ids(values: list[Any]) -> list[str]:
    out: list[str] = []
    for value in values:
        if not isinstance(value, str):
            continue
        item = value.strip()
        if item and item not in out:
            out.append(item)
    return out


def _baseline_active_ids(extras: dict[str, Any]) -> list[str]:
    block = extras.get("baseline_modules")
    if not isinstance(block, dict) or "active" not in block:
        return list(PROFILE_BASELINE_MODULES)
    raw = block.get("active")
    if not isinstance(raw, list):
        return []
    allowed = set(PROFILE_BASELINE_MODULES)
    return _unique_ids([item for item in raw if isinstance(item, str) and item in allowed])


def _ondemand_active_ids(extras: dict[str, Any]) -> list[str]:
    block = extras.get("ondemand_modules")
    if not isinstance(block, dict):
        return []
    raw = block.get("active") or []
    if not isinstance(raw, list):
        return []
    return _unique_ids([item for item in raw if isinstance(item, str)])


def subscriptions_for_active_modules(
    *,
    baseline_active: list[str] | None = None,
    ondemand_active: list[str] | None = None,
) -> list[dict[str, str]]:
    ids = _unique_ids(list(baseline_active or []) + list(ondemand_active or []))
    seen: set[tuple[str, str]] = set()
    out: list[dict[str, str]] = []
    for module_id in ids:
        for item in MODULE_HOOK_SUBSCRIPTIONS.get(module_id, ()):
            key = (item.get("event", ""), item.get("module", ""))
            if key in seen:
                continue
            seen.add(key)
            out.append(dict(item))
    return out


def sync_event_bridge_subscriptions(extras: dict[str, Any]) -> dict[str, Any]:
    """Fill or clear `event_bridge.subscriptions` from still-active modules."""
    bridge = extras.get("event_bridge")
    if not isinstance(bridge, dict):
        bridge = default_event_bridge()
        extras["event_bridge"] = bridge
    bridge["subscriptions"] = subscriptions_for_active_modules(
        baseline_active=_baseline_active_ids(extras),
        ondemand_active=_ondemand_active_ids(extras),
    )
    return extras


def _default_readiness() -> dict[str, Any]:
    return {
        provider_id: {
            "status": "unknown",
            "capability": SHIPPED_PROVIDER_CAPABILITY[provider_id],
            "evidence": None,
        }
        for provider_id in SHIPPED_MIDDLEWARE_PROVIDERS
    }


def default_middleware_providers() -> dict[str, Any]:
    return {
        "schema_version": STAGE6_SCHEMA_VERSION,
        "source": STAGE6_SOURCE,
        "registered": list(SHIPPED_MIDDLEWARE_PROVIDERS),
        "required": list(DEFAULT_REQUIRED_MIDDLEWARE_PROVIDERS),
        "active": [],
        "degraded": [],
        "readiness": _default_readiness(),
    }


def default_capability_router() -> dict[str, Any]:
    return {
        "schema_version": STAGE6_SCHEMA_VERSION,
        "source": STAGE6_SOURCE,
        "exact": True,
        "semantic": True,
        "structural": True,
        "external": True,
    }


def dispatch_hook_event(
    event_bridge: dict[str, Any],
    event: str,
    *,
    source: str = "cursor-hooks",
    at: str | None = None,
) -> dict[str, Any]:
    name = (event or "").strip()
    if not name:
        raise AdapterMiddlewareError("event_bridge.last_event.event must be a non-empty string")
    delivered: list[str] = []
    skipped: list[str] = []
    for item in event_bridge.get("subscriptions") or []:
        if not isinstance(item, dict):
            continue
        module = str(item.get("module") or "").strip()
        if not module:
            continue
        if item.get("event") == name:
            if module not in delivered:
                delivered.append(module)
        else:
            skipped.append(module)
    return {
        "event": name,
        "at": at or "",
        "source": source,
        "delivered": delivered,
        "skipped": skipped,
    }


def classify_transport_probe(
    *,
    present: bool | None = None,
    reachable: bool | None = None,
    available: bool | None = None,
    status: str | None = None,
    evidence: str | None = None,
) -> dict[str, Any]:
    if status in {"ready", "missing", "failed", "unknown"}:
        return {"status": status, "evidence": evidence}
    resolved_present = present if present is not None else available
    if resolved_present is False:
        return {"status": "missing", "evidence": evidence}
    if resolved_present is True and reachable is False:
        return {"status": "failed", "evidence": evidence}
    if resolved_present is True:
        return {"status": "ready", "evidence": evidence}
    return {"status": "unknown", "evidence": evidence}


def mcp_server_ids_from_config(raw: Any) -> list[str]:
    if not isinstance(raw, dict):
        return []
    servers = raw.get("mcpServers")
    if not isinstance(servers, dict):
        return []
    return [str(name) for name in servers.keys()]


def select_registered_mcp_servers(
    configured: list[str] | tuple[str, ...],
    registered: list[str] | tuple[str, ...] | None = None,
) -> list[str]:
    allow = list(registered if registered is not None else SHIPPED_MIDDLEWARE_PROVIDERS)
    out: list[str] = []
    for name in configured:
        if name in allow and name not in out:
            out.append(name)
    return out


def probe_shipped_provider_readiness(
    provider_id: str,
    *,
    present: bool | None = None,
    reachable: bool | None = None,
    available: bool | None = None,
    status: str | None = None,
    evidence: str | None = None,
) -> dict[str, Any]:
    classified = classify_transport_probe(
        present=present,
        reachable=reachable,
        available=available,
        status=status,
        evidence=evidence,
    )
    return {
        "status": classified["status"],
        "capability": SHIPPED_PROVIDER_CAPABILITY.get(
            provider_id,
            EXTERNAL_KNOWLEDGE_CAPABILITY if provider_id == SMART_SEARCH_PROVIDER else "unknown",
        ),
        "evidence": classified["evidence"],
    }


def apply_middleware_probes(
    providers: dict[str, Any],
    probes: dict[str, Any],
) -> dict[str, Any]:
    registered = providers.get("registered") or []
    readiness = dict(providers.get("readiness") or {})
    if not isinstance(probes, dict):
        return providers
    for provider_id, value in probes.items():
        if provider_id not in registered or not isinstance(value, dict):
            continue
        readiness[provider_id] = probe_shipped_provider_readiness(
            provider_id,
            present=value.get("present"),
            reachable=value.get("reachable"),
            available=value.get("available"),
            status=value.get("status"),
            evidence=value.get("evidence") if isinstance(value.get("evidence"), str) else None,
        )
    next_providers = {**providers, "readiness": readiness}
    return next_providers


def probe_smart_search_readiness(*, available: bool | None = None, status: str | None = None) -> dict[str, Any]:
    return probe_shipped_provider_readiness(
        SMART_SEARCH_PROVIDER,
        available=available,
        status=status,
    )


def normalize_capability_router(raw: Any) -> dict[str, Any]:
    if raw is None:
        return default_capability_router()
    if not isinstance(raw, dict):
        raise AdapterMiddlewareError("capability_router must be a JSON object")
    for key in list(raw.keys()):
        if key in OPTIONAL_CODE_INTEL_PROVIDERS or key in {
            "fast_context_search",
            "codegraph_explore",
            "codegraph_search",
            "codegraph_callers",
        }:
            raise AdapterMiddlewareError("capability_router must not bind Optional tool names")
    next_router = default_capability_router()
    for intent in RETRIEVAL_INTENTS:
        if raw.get(intent) is False:
            next_router.pop(intent, None)
        else:
            next_router[intent] = True
    return next_router


def required_capabilities(extras: dict[str, Any]) -> list[str]:
    raw = extras.get("required_capabilities") or []
    if not isinstance(raw, list):
        return []
    out: list[str] = []
    for item in raw:
        if isinstance(item, str) and item.strip() and item not in out:
            out.append(item.strip())
    return out


def assert_external_knowledge(
    extras: dict[str, Any],
    readiness: dict[str, Any],
    *,
    phase: str,
) -> None:
    if phase in {"create", "patch"}:
        return
    if EXTERNAL_KNOWLEDGE_CAPABILITY not in required_capabilities(extras):
        return
    if readiness.get("status") == "ready":
        return
    if extras.get("external_knowledge_policy") == "degrade":
        extras["profile_health"] = "degraded"
        return
    raise AdapterMiddlewareError(
        "external-knowledge required but smart-search Provider is not ready"
    )

File: packages/core/src/task/adapter-middleware.ts
/**
 * Stage 6 Cursor Adapter + Middleware (P27 Middleware, P29 Adapter/Profile,
 * P30 Stage 6).
 *
 * Pure helpers persisted through existing create extras / patch. Does not
 * write Kernel store or task.json. Field names here are an implementation
 * choice — not a frozen Manifest / Event ABI schema.
 */

import { KernelError } from "./kernel-contract.js";
import { isPlainObject } from "./schema.js";

export const STAGE6_SOURCE = "stage6-adapter-middleware";
export const STAGE6_SCHEMA_VERSION = 1 as const;

export const STAGE6_HOOK_EVENTS = [
  "sessionStart",
  "preToolUse",
  "beforeSubmitPrompt",
  "beforeShellExecution",
  "afterShellExecution",
  "stop",
] as const;

export type Stage6HookEvent = (typeof STAGE6_HOOK_EVENTS)[number];

export const RETRIEVAL_INTENTS = [
  "exact",
  "semantic",
  "structural",
  "external",
] as const;

export type RetrievalIntent = (typeof RETRIEVAL_INTENTS)[number];

export const DEFAULT_MIDDLEWARE_PROVIDERS = ["smart-search"] as const;
export const OPTIONAL_CODE_INTEL_PROVIDERS = [
  "codegraph",
  "fast-context",
] as const;

export const SMART_SEARCH_PROVIDER = "smart-search";
export const EXTERNAL_KNOWLEDGE_CAPABILITY = "external-knowledge";

export type Stage6CommandPhase = "create" | "start" | "archive" | "patch";

export type ProviderReadinessStatus =
  | "ready"
  | "missing"
  | "failed"
  | "unknown";

export interface EventSubscription {
  event: string;
  module: string;
}

export interface EventBridgeLastEvent {
  event: string;
  at: string;
  source: string;
  delivered: string[];
  skipped: string[];
}

export interface EventBridgeState {
  schema_version: typeof STAGE6_SCHEMA_VERSION;
  source: typeof STAGE6_SOURCE;
  subscriptions: EventSubscription[];
  last_event?: EventBridgeLastEvent;
}

export interface ProviderReadiness {
  status: ProviderReadinessStatus;
  capability: string;
  evidence?: string | null;
}

export interface MiddlewareProviders {
  schema_version: typeof STAGE6_SCHEMA_VERSION;
  source: typeof STAGE6_SOURCE;
  registered: string[];
  required: string[];
  active: string[];
  degraded: string[];
  readiness: Record<string, ProviderReadiness>;
}

export interface CapabilityRouter {
  schema_version: typeof STAGE6_SCHEMA_VERSION;
  source: typeof STAGE6_SOURCE;
  exact?: true;
  semantic?: true;
  structural?: true;
  external?: true;
}

export interface DispatchHookEventInput {
  event: string;
  source?: string;
  at?: string;
}

export function defaultEventBridge(): EventBridgeState {
  return {
    schema_version: STAGE6_SCHEMA_VERSION,
    source: STAGE6_SOURCE,
    subscriptions: [],
  };
}

export function defaultMiddlewareProviders(): MiddlewareProviders {
  return {
    schema_version: STAGE6_SCHEMA_VERSION,
    source: STAGE6_SOURCE,
    registered: [...DEFAULT_MIDDLEWARE_PROVIDERS],
    required: [...DEFAULT_MIDDLEWARE_PROVIDERS],
    active: [],
    degraded: [],
    readiness: {
      [SMART_SEARCH_PROVIDER]: {
        status: "unknown",
        capability: EXTERNAL_KNOWLEDGE_CAPABILITY,
        evidence: null,
      },
    },
  };
}

export function defaultCapabilityRouter(): CapabilityRouter {
  return {
    schema_version: STAGE6_SCHEMA_VERSION,
    source: STAGE6_SOURCE,
    exact: true,
    semantic: true,
    structural: true,
    external: true,
  };
}

export function subscribeEvent(
  extras: Record<string, unknown>,
  event: string,
  moduleName: string,
): EventBridgeState {
  const current = readEventBridge(extras);
  const eventName = requireId(event, "event_bridge.subscriptions[].event");
  const module = requireId(moduleName, "event_bridge.subscriptions[].module");
  const subscriptions = [...current.subscriptions];
  if (!subscriptions.some((item) => item.event === eventName && item.module === module)) {
    subscriptions.push({ event: eventName, module });
  }
  const next: EventBridgeState = { ...current, subscriptions };
  extras.event_bridge = next;
  return next;
}

/**
 * Record and dispatch one hook event. Unsubscribed modules are skipped and
 * never fail the hook.
 */
export function recordHookEvent(
  extras: Record<string, unknown>,
  input: DispatchHookEventInput,
): EventBridgeLastEvent {
  const current = readEventBridge(extras);
  const event = requireId(input.event, "event_bridge.last_event.event");
  const delivered: string[] = [];
  const skipped: string[] = [];
  for (const sub of current.subscriptions) {
    if (sub.event !== event) {
      skipped.push(sub.module);
      continue;
    }
    if (!delivered.includes(sub.module)) delivered.push(sub.module);
  }
  const lastEvent: EventBridgeLastEvent = {
    event,
    at: typeof input.at === "string" && input.at.trim() !== ""
      ? input.at
      : new Date().toISOString(),
    source:
      typeof input.source === "string" && input.source.trim() !== ""
        ? input.source
        : "cursor-hooks",
    delivered,
    skipped,
  };
  extras.event_bridge = { ...current, last_event: lastEvent };
  return lastEvent;
}

export function probeSmartSearchReadiness(input: {
  available?: boolean;
  status?: ProviderReadinessStatus;
  evidence?: string | null;
} = {}): ProviderReadiness {
  const status =
    input.status ??
    (input.available === true
      ? "ready"
      : input.available === false
        ? "missing"
        : "unknown");
  return {
    status,
    capability: EXTERNAL_KNOWLEDGE_CAPABILITY,
    evidence: input.evidence ?? null,
  };
}

export function applySmartSearchReadiness(
  extras: Record<string, unknown>,
  readiness: ProviderReadiness,
): MiddlewareProviders {
  const current = readMiddlewareProviders(extras);
  const next: MiddlewareProviders = {
    ...current,
    readiness: {
      ...current.readiness,
      [SMART_SEARCH_PROVIDER]: {
        ...readiness,
        capability: EXTERNAL_KNOWLEDGE_CAPABILITY,
      },
    },
  };
  extras.middleware_providers = applyProviderHealth(next, extras);
  return extras.middleware_providers as MiddlewareProviders;
}

export function normalizeEventBridge(raw: unknown): EventBridgeState {
  if (raw === undefined || raw === null) return defaultEventBridge();
  if (!isPlainObject(raw)) {
    throw new KernelError("INVALID_REQUEST", "event_bridge must be a JSON object");
  }
  const subscriptionsIn = raw.subscriptions;
  const subscriptions: EventSubscription[] = [];
  if (subscriptionsIn !== undefined && subscriptionsIn !== null) {
    if (!Array.isArray(subscriptionsIn)) {
      throw new KernelError(
        "INVALID_REQUEST",
        "event_bridge.subscriptions must be an array",
      );
    }
    for (const item of subscriptionsIn) {
      if (!isPlainObject(item)) {
        throw new KernelError(
          "INVALID_REQUEST",
          "event_bridge.subscriptions[] must be objects",
        );
      }
      const sub: EventSubscription = {
        event: requireId(item.event, "event_bridge.subscriptions[].event"),
        module: requireId(item.module, "event_bridge.subscriptions[].module"),
      };
      if (
        !subscriptions.some(
          (existing) =>
            existing.event === sub.event && existing.module === sub.module,
        )
      ) {
        subscriptions.push(sub);
      }
    }
  }
  const lastRaw = raw.last_event;
  let lastEvent: EventBridgeLastEvent | undefined;
  if (lastRaw !== undefined && lastRaw !== null) {
    if (!isPlainObject(lastRaw)) {
      throw new KernelError(
        "INVALID_REQUEST",
        "event_bridge.last_event must be an object",
      );
    }
    lastEvent = {
      event: requireId(lastRaw.event, "event_bridge.last_event.event"),
      at:
        typeof lastRaw.at === "string" && lastRaw.at.trim() !== ""
          ? lastRaw.at
          : new Date().toISOString(),
      source:
        typeof lastRaw.source === "string" && lastRaw.source.trim() !== ""
          ? lastRaw.source
          : "cursor-hooks",
      delivered: uniqueStrings(asStringArray(lastRaw.delivered)),
      skipped: uniqueStrings(asStringArray(lastRaw.skipped)),
    };
  }
  return {
    schema_version: STAGE6_SCHEMA_VERSION,
    source: STAGE6_SOURCE,
    subscriptions,
    ...(lastEvent ? { last_event: lastEvent } : {}),
  };
}

export function normalizeMiddlewareProviders(raw: unknown): MiddlewareProviders {
  if (raw === undefined || raw === null) return defaultMiddlewareProviders();
  if (!isPlainObject(raw)) {
    throw new KernelError(
      "INVALID_REQUEST",
      "middleware_providers must be a JSON object",
    );
  }
  const defaults = defaultMiddlewareProviders();
  const registered = uniqueStrings([
    ...defaults.registered,
    ...asStringArray(raw.registered),
  ]);
  const required = uniqueStrings(asStringArray(raw.required)).filter((name) =>
    registered.includes(name),
  );
  const active = uniqueStrings(asStringArray(raw.active)).filter((name) =>
    registered.includes(name),
  );
  const degraded = uniqueStrings(asStringArray(raw.degraded));
  const readiness: Record<string, ProviderReadiness> = {
    ...defaults.readiness,
  };
  if (isPlainObject(raw.readiness)) {
    for (const [name, value] of Object.entries(raw.readiness)) {
      if (!isPlainObject(value)) continue;
      const status = parseReadinessStatus(value.status);
      readiness[name] = {
        status,
        capability:
          typeof value.capability === "string" && value.capability.trim() !== ""
            ? value.capability
            : name === SMART_SEARCH_PROVIDER
              ? EXTERNAL_KNOWLEDGE_CAPABILITY
              : "unknown",
        evidence:
          typeof value.evidence === "string" ? value.evidence : null,
      };
    }
  }
  return {
    schema_version: STAGE6_SCHEMA_VERSION,
    source: STAGE6_SOURCE,
    registered,
    required: required.length > 0 ? required : [...defaults.required],
    active,
    degraded,
    readiness,
  };
}

export function normalizeCapabilityRouter(raw: unknown): CapabilityRouter {
  if (raw === undefined || raw === null) return defaultCapabilityRouter();
  if (!isPlainObject(raw)) {
    throw new KernelError(
      "INVALID_REQUEST",
      "capability_router must be a JSON object",
    );
  }
  const forbidden = [
    ...OPTIONAL_CODE_INTEL_PROVIDERS,
    "fast_context_search",
    "codegraph_explore",
    "codegraph_search",
    "codegraph_callers",
  ];
  for (const key of Object.keys(raw)) {
    if (forbidden.includes(key)) {
      throw new KernelError(
        "INVALID_REQUEST",
        "capability_router must not bind Optional tool names",
      );
    }
  }
  const next: CapabilityRouter = {
    schema_version: STAGE6_SCHEMA_VERSION,
    source: STAGE6_SOURCE,
  };
  for (const intent of RETRIEVAL_INTENTS) {
    if (raw[intent] !== false) {
      next[intent] = true;
    }
  }
  return next;
}

export function readEventBridge(extras: Record<string, unknown>): EventBridgeState {
  return normalizeEventBridge(extras.event_bridge);
}

export function readMiddlewareProviders(
  extras: Record<string, unknown>,
): MiddlewareProviders {
  return normalizeMiddlewareProviders(extras.middleware_providers);
}

export function readCapabilityRouter(
  extras: Record<string, unknown>,
): CapabilityRouter {
  return normalizeCapabilityRouter(extras.capability_router);
}

export function requiredCapabilities(extras: Record<string, unknown>): string[] {
  return uniqueStrings(asStringArray(extras.required_capabilities));
}

export function externalKnowledgeReady(providers: MiddlewareProviders): boolean {
  const readiness = providers.readiness[SMART_SEARCH_PROVIDER];
  return readiness?.status === "ready";
}

export function normalizeStage6InExtras(extras: Record<string, unknown>): void {
  extras.event_bridge = normalizeEventBridge(extras.event_bridge);
  extras.capability_router = normalizeCapabilityRouter(extras.capability_router);
  const incoming = extras.hook_event;
  if (isPlainObject(incoming) && typeof incoming.event === "string") {
    recordHookEvent(extras, {
      event: incoming.event,
      source: typeof incoming.source === "string" ? incoming.source : undefined,
      at: typeof incoming.at === "string" ? incoming.at : undefined,
    });
    delete extras.hook_event;
  }
  extras.middleware_providers = normalizeMiddlewareProviders(
    extras.middleware_providers,
  );
  if (isPlainObject(extras.smart_search_probe)) {
    const probeStatus = extras.smart_search_probe.status;
    applySmartSearchReadiness(
      extras,
      probeSmartSearchReadiness({
        available:
          extras.smart_search_probe.available === true
            ? true
            : extras.smart_search_probe.available === false
              ? false
              : undefined,
        status:
          probeStatus === "ready" ||
          probeStatus === "missing" ||
          probeStatus === "failed" ||
          probeStatus === "unknown"
            ? probeStatus
            : undefined,
        evidence:
          typeof extras.smart_search_probe.evidence === "string"
            ? extras.smart_search_probe.evidence
            : null,
      }),
    );
    delete extras.smart_search_probe;
  } else {
    extras.middleware_providers = applyProviderHealth(
      extras.middleware_providers as MiddlewareProviders,
      extras,
    );
  }
}

export function assertStage6ForPhase(
  extras: Record<string, unknown>,
  phase: Stage6CommandPhase,
): void {
  if (phase === "create" || phase === "patch") return;
  const providers = readMiddlewareProviders(extras);
  const needsExternal = requiredCapabilities(extras).includes(
    EXTERNAL_KNOWLEDGE_CAPABILITY,
  );
  if (!needsExternal) return;
  if (externalKnowledgeReady(providers)) return;
  if (extras.external_knowledge_policy === "degrade") {
    extras.profile_health = "degraded";
    return;
  }
  throw new KernelError(
    "INVALID_TRANSITION",
    "external-knowledge required but smart-search Provider is not ready",
  );
}

export function normalizeStage6InExtrasAndAssert(
  extras: Record<string, unknown>,
  phase: Stage6CommandPhase,
): void {
  normalizeStage6InExtras(extras);
  assertStage6ForPhase(extras, phase);
}

function applyProviderHealth(
  providers: MiddlewareProviders,
  extras: Record<string, unknown>,
): MiddlewareProviders {
  const degraded = uniqueStrings([...providers.degraded]);
  const smart = providers.readiness[SMART_SEARCH_PROVIDER];
  if (smart && smart.status !== "ready" && smart.status !== "unknown") {
    if (!degraded.includes(SMART_SEARCH_PROVIDER)) {
      degraded.push(SMART_SEARCH_PROVIDER);
    }
    extras.profile_health = "degraded";
  }
  const missing = uniqueStrings(asStringArray(extras.ondemand_required_missing));
  if (
    missing.includes(EXTERNAL_KNOWLEDGE_CAPABILITY) ||
    missing.includes(SMART_SEARCH_PROVIDER)
  ) {
    if (!degraded.includes(SMART_SEARCH_PROVIDER)) {
      degraded.push(SMART_SEARCH_PROVIDER);
    }
    extras.profile_health = "degraded";
  }
  return { ...providers, degraded };
}

function parseReadinessStatus(value: unknown): ProviderReadinessStatus {
  if (
    value === "ready" ||
    value === "missing" ||
    value === "failed" ||
    value === "unknown"
  ) {
    return value;
  }
  return "unknown";
}

function asStringArray(value: unknown): string[] {
  if (!Array.isArray(value)) return [];
  return value.filter((item): item is string => typeof item === "string");
}

function uniqueStrings(values: readonly unknown[]): string[] {
  const out: string[] = [];
  for (const value of values) {
    if (typeof value !== "string") continue;
    const id = value.trim();
    if (!id || out.includes(id)) continue;
    out.push(id);
  }
  return out;
}

function requireId(value: unknown, field: string): string {
  if (typeof value !== "string" || value.trim() === "") {
    throw new KernelError("INVALID_REQUEST", `${field} must be a non-empty string`);
  }
  return value.trim();
}
